File: ai-qa-automation-platform/apps/api/app/exec/orchestrator.py
"""Execution orchestration — run a workspace's approved test suite.

Flow: compile suites → execute (worker pool) → eval (DeepEval) → gate
verdict → on failure: defect report + closed-loop review items.
"""
from __future__ import annotations

import uuid
from typing import Any

from sqlalchemy.orm import Session

from ..agents.phase_agents import ReleaseGaterAgent
from ..db.models import (
    MetricScore,
    ReviewItem,
    Run,
    RunResult,
    TestCase,
    WorkspaceSettings,
)
from ..eval.metrics import EvalInput, evaluate_case
from ..observability.langfuse import tracer

TEST_TYPE_TO_METRICS: dict[str, list[str]] = {
    "ui": ["answer_relevancy", "groundedness", "tool_sequence_accuracy"],
    "api": ["answer_relevancy", "completeness", "correctness"],
    "db": ["groundedness", "completeness"],
    "integration": ["tool_sequence_accuracy", "correctness"],
}


def create_run(db: Session, workspace_id: str, trigger: str = "manual") -> Run:
    """Create a Run record with a threshold snapshot from settings."""
    s = db.get(WorkspaceSettings, workspace_id)
    thresholds = s.thresholds if s else {}
    run = Run(
        id=str(uuid.uuid4()),
        workspace_id=workspace_id,
        trigger=trigger,
        status="queued",
        threshold_snapshot=thresholds,
    )
    db.add(run)
    db.commit()
    return run


def execute_run(db: Session, run: Run) -> dict[str, Any]:
    """Execute all approved test cases for the run's workspace."""
    run.status = "running"
    db.commit()
    trace = tracer.start_run_trace(run.workspace_id, run.id)

    cases = (
        db.query(TestCase)
        .filter(
            TestCase.workspace_id == run.workspace_id,
            TestCase.status == "approved",
        )
        .all()
    )

    results_summary: dict[str, Any] = {"passed": 0, "failed": 0, "total": len(cases)}
    metric_rows: list[MetricScore] = []

    for case in cases:
        outcome = _run_case(case)
        result = RunResult(
            id=str(uuid.uuid4()),
            run_id=run.id,
            test_case_id=case.id,
            workspace_id=run.workspace_id,
            status=outcome["status"],
            duration_ms=outcome["duration_ms"],
            error=outcome["error"],
        )
        db.add(result)
        db.flush()
        if outcome["status"] == "passed":
            results_summary["passed"] += 1
        else:
            results_summary["failed"] += 1

        # Eval per test type (only when execution passed — else it's a defect)
        if outcome["status"] == "passed":
            eval_input = EvalInput(
                input=case.title,
                actual_output=outcome.get("output", case.code or ""),
                retrieval_context=[case.derived_from or ""],
                tools_called=outcome.get("tools_called", []),
                expected_tools=outcome.get("expected_tools", []),
                high_risk=case.priority == "P0",
            )
            for mr in evaluate_case(eval_input, run.threshold_snapshot):
                if mr.metric not in TEST_TYPE_TO_METRICS.get(case.test_type, []):
                    continue
                metric_rows.append(
                    MetricScore(
                        id=str(uuid.uuid4()),
                        run_result_id=result.id,
                        workspace_id=run.workspace_id,
                        metric=mr.metric,
                        score=mr.score,
                        threshold=mr.threshold,
                        hard_gate=mr.hard_gate,
                        passed=mr.passed,
                    )
                )

    db.add_all(metric_rows)
    db.commit()

    # Gate verdict
    gate = _gate(db, run)
    run.status = "passed" if gate["verdict"] == "pass" else "blocked"
    run.gate_verdict = gate["verdict"]
    db.commit()
    tracer.flush()

    # Closed loop: failures → review items
    for result in db.query(RunResult).filter(RunResult.run_id == run.id, RunResult.status != "passed"):
        db.add(
            ReviewItem(
                id=str(uuid.uuid4()),
                workspace_id=run.workspace_id,
                kind="defect",
                test_case_id=result.test_case_id,
                run_result_id=result.id,
                reason=result.error or "failed in run",
            )
        )
    db.commit()

    return {"run_id": run.id, "gate": gate, **results_summary}


def _run_case(case: TestCase) -> dict[str, Any]:
    """Execute one case in-process (worker pool integration lands in M5).

    Local stub: a case whose code contains 'def test_' is treated as passing;
    otherwise it reports a failure. Real Playwright/pytest workers replace this.
    """
    code = case.code or ""
    if "def test_" in code:
        return {"status": "passed", "duration_ms": 120, "output": "ok"}
    return {"status": "failed", "duration_ms": 40, "error": "no runnable code"}


def _gate(db: Session, run: Run) -> dict[str, Any]:
    """Release Gater agent decides pass/block from metric scores.

    Fail-closed: if the LLM verdict is unavailable (no provider/key, retry
    status), fall back to a deterministic hard-gate check so a broken LLM
    never lets a release through.
    """
    scores = (
        db.query(MetricScore)
        .join(RunResult, RunResult.id == MetricScore.run_result_id)
        .filter(RunResult.run_id == run.id)
        .all()
    )
    score_data = [
        {"metric": s.metric, "score": s.score, "threshold": s.threshold, "hard_gate": s.hard_gate}
        for s in scores
    ]
    from ..agents.base import AgentContext

    ctx = AgentContext(
        workspace_id=run.workspace_id,
        data={"metric_scores": score_data, "thresholds": run.threshold_snapshot},
    )
    result = ReleaseGaterAgent().run(ctx)

    verdict = result.output.get("verdict")
    if verdict not in {"pass", "block"}:
        # Deterministic fallback: any failed hard-gate metric blocks.
        blocked = [s["metric"] for s in score_data if s["hard_gate"] and s["score"] < s["threshold"]]
        verdict = "block" if blocked else "pass"
        return {"verdict": verdict, "blocked_metrics": blocked}

    blocked_metrics = result.output.get("blocked_metrics", [])
    if verdict == "block" and not blocked_metrics:
        blocked_metrics = [
            s["metric"] for s in score_data if s["hard_gate"] and s["score"] < s["threshold"]
        ]
    return {"verdict": verdict, "blocked_metrics": blocked_metrics}

File: ai-qa-automation-platform/apps/web/lib/exec/orchestrator.ts
import { prisma } from "../db";
import { evaluateCase, MetricName, HARD_GATE_METRICS } from "../eval/metrics";
import { ReleaseGaterAgent } from "../agents/phase-agents";
import type { AgentContext } from "../agents/base";

/**
 * Run orchestrator — execute → eval → gate → closed-loop review items.
 * Serverless-safe: all work happens in the request (or Vercel cron).
 */
const TEST_TYPE_TO_METRICS: Record<string, MetricName[]> = {
  ui: ["answer_relevancy", "groundedness", "tool_sequence_accuracy"],
  api: ["answer_relevancy", "completeness", "correctness"],
  db: ["groundedness", "completeness"],
  integration: ["tool_sequence_accuracy", "correctness"],
};

export async function createRun(workspaceId: string, trigger = "manual") {
  const settings = await prisma.workspaceSettings.findUnique({ where: { workspaceId } });
  const thresholds = (settings?.thresholds as Record<string, number> | null) ?? {};
  return prisma.run.create({
    data: {
      workspaceId,
      trigger,
      status: "queued",
      thresholdSnapshot: thresholds,
    },
  });
}

export async function executeRun(runId: string) {
  const run = await prisma.run.findUnique({ where: { id: runId }, include: { workspace: true } });
  if (!run) throw new Error("Run not found");

  await prisma.run.update({ where: { id: runId }, data: { status: "running" } });

  const cases = await prisma.testCase.findMany({
    where: { workspaceId: run.workspaceId, status: "approved" },
  });

  let passed = 0;
  let failed = 0;
  const metricRows = [];

  for (const caseItem of cases) {
    const outcome = runCase(caseItem.code ?? "");
    const result = await prisma.runResult.create({
      data: {
        runId,
        testCaseId: caseItem.id,
        workspaceId: run.workspaceId,
        status: outcome.status,
        durationMs: outcome.durationMs,
        error: outcome.error,
      },
    });

    if (outcome.status === "passed") passed++;
    else failed++;

    if (outcome.status === "passed") {
      const metricResults = await evaluateCase(
        {
          input: caseItem.title,
          actualOutput: outcome.output ?? caseItem.code ?? "",
          retrievalContext: [caseItem.derivedFrom ?? ""],
          toolsCalled: outcome.toolsCalled ?? [],
          expectedTools: outcome.expectedTools ?? [],
          highRisk: caseItem.priority === "P0",
        },
        (run.thresholdSnapshot as Record<string, number>) ?? {},
      );
      const allowed = TEST_TYPE_TO_METRICS[caseItem.testType] ?? [];
      for (const mr of metricResults) {
        if (!allowed.includes(mr.metric)) continue;
        metricRows.push(
          prisma.metricScore.create({
            data: {
              runResultId: result.id,
              workspaceId: run.workspaceId,
              metric: mr.metric,
              score: mr.score,
              threshold: mr.threshold,
              hardGate: mr.hardGate,
              passed: mr.passed,
            },
          }),
        );
      }
    }
  }

  await Promise.all(metricRows);

  // Gate verdict
  const scores = await prisma.metricScore.findMany({
    where: { runResult: { runId }, workspaceId: run.workspaceId },
  });
  const scoreData = scores.map((s) => ({
    metric: s.metric,
    score: s.score,
    threshold: s.threshold,
    hard_gate: s.hardGate,
  }));

  const ctx: AgentContext = {
    workspaceId: run.workspaceId,
    data: { metricScores: scoreData, thresholds: run.thresholdSnapshot },
  };
  const gater = new ReleaseGaterAgent();
  const gateResult = await gater.run(ctx);

  let verdict: string | null = gateResult.output.verdict as string;
  let blockedMetrics: string[] = (gateResult.output.blocked_metrics as string[]) ?? [];
  if (verdict !== "pass" && verdict !== "block") {
    // Fail-closed deterministic fallback
    blockedMetrics = scoreData
      .filter((s) => HARD_GATE_METRICS.includes(s.metric as MetricName) && s.score < s.threshold)
      .map((s) => s.metric);
    verdict = blockedMetrics.length ? "block" : "pass";
  }
  if (verdict === "block" && !blockedMetrics.length) {
    blockedMetrics = scoreData
      .filter((s) => HARD_GATE_METRICS.includes(s.metric as MetricName) && s.score < s.threshold)
      .map((s) => s.metric);
  }

  await prisma.run.update({
    where: { id: runId },
    data: { status: verdict === "pass" ? "passed" : "blocked", gateVerdict: verdict },
  });

  // Closed loop: failures → review items
  const failedResults = await prisma.runResult.findMany({
    where: { runId, status: { not: "passed" } },
  });
  for (const r of failedResults) {
    await prisma.reviewItem.create({
      data: {
        workspaceId: run.workspaceId,
        kind: "defect",
        testCaseId: r.testCaseId,
        runResultId: r.id,
        reason: r.error ?? "failed in run",
      },
    });
  }

  return { runId, gate: { verdict, blocked_metrics: blockedMetrics }, passed, failed, total: cases.length };
}

/** In-process case execution — real Playwright/pytest workers land later. */
function runCase(code: string): { status: string; durationMs: number; error?: string; output?: string; toolsCalled?: string[]; expectedTools?: string[] } {
  if (code && code.includes("def test_")) {
    return { status: "passed", durationMs: 120, output: "ok", toolsCalled: ["run_test"], expectedTools: ["run_test"] };
  }
  return { status: "failed", durationMs: 40, error: "no runnable code" };
}
